### PR TITLE
depth-clip-control feature is unimplemented

### PR DESCRIPTION
**Pull request by mwyrzykowski**
#### 1a696df1b0a026a95c9808f0307758a8a8c268d1
<pre>
depth-clip-control feature is unimplemented
<a href="https://bugs.webkit.org/show_bug.cgi?id=251664">https://bugs.webkit.org/show_bug.cgi?id=251664</a>
&lt;radar://104992830&gt;

Reviewed by Myles C. Maxfield.

Implement depth clip control which controls if fragments
with depth values outside of [0, 1] are clipped or not.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::convertToBacking):

* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::baseFeatures):
All devices support depth clamping.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setStencilReference):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::depthClipMode const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
Pass the feature through to the command encoder.

* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:
Update two-cubes to illustrate this feature.

Canonical link: <a href="https://commits.webkit.org/260723@main">https://commits.webkit.org/260723@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/664a4909257e2bf6ed24acde69c2c08b78b222c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41828 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/542 "Reverted pull request changes (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118284 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9385 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101244 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114765 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97893 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42790 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96625 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29535 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84539 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10871 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30883 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7814 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16995 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50481 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7396 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13214 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebGPU/WebGPU/RenderPipeline.h [hunk 1]
```diff
@@ -43,14 +43,14 @@ class PipelineLayout;
 class RenderPipeline : public WGPURenderPipelineImpl, public RefCounted<RenderPipeline> {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    static Ref<RenderPipeline> create(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, MTLDepthStencilDescriptor *depthStencilDescriptor, MTLRenderPipelineReflection *reflection, Device& device)
+    static Ref<RenderPipeline> create(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, MTLDepthClipMode depthClipMode, MTLDepthStencilDescriptor *depthStencilDescriptor, MTLRenderPipelineReflection *reflection, Device& device)
```
Comment (djg) [inline Source/WebGPU/WebGPU/RenderPipeline.h]: Is it preferable to keep adding arguments to create calls or should we pass arguments via a descriptor struct?

**Review by djg — APPROVED** [review]
:shipit:

File: Source/WebGPU/WebGPU/RenderPipeline.h (hunk 1 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/RenderPipeline.h]: if `RenderPipeline::create` was more widely used I would say descriptor struct for sure but let me remove one of the overloads since I think that is not needed

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp [hunk 2]
```diff
@@ -474,7 +481,7 @@ static auto convertToBacking(const RenderPipelineDescriptor& descriptor, Convert
             static_cast<uint32_t>(backingBuffers.size()),
             backingBuffers.data(),
         }, {
-            nullptr,
+            descriptor.primitive && descriptor.primitive->unclippedDepth ? &depthClipControl.chain : nullptr,
```
Comment (litherum) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:484]: We should only do this if the clip control feature is enabled.

File: Source/WebGPU/WebGPU/RenderPipeline.h [hunk 3]
```diff
@@ -43,14 +43,9 @@ class PipelineLayout;
 class RenderPipeline : public WGPURenderPipelineImpl, public RefCounted<RenderPipeline> {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    static Ref<RenderPipeline> create(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, MTLDepthStencilDescriptor *depthStencilDescriptor, MTLRenderPipelineReflection *reflection, Device& device)
+    static Ref<RenderPipeline> create(id<MTLRenderPipelineState> renderPipelineState, MTLPrimitiveType primitiveType, std::optional<MTLIndexType> indexType, MTLWinding frontFace, MTLCullMode cullMode, MTLDepthClipMode depthClipMode, MTLDepthStencilDescriptor *depthStencilDescriptor, MTLRenderPipelineReflection *reflection, const PipelineLayout *pipelineLayout, Device& device)
```
Comment (litherum) [inline Source/WebGPU/WebGPU/RenderPipeline.h:46]: This is getting pretty unwieldy, can we batch these together in a struct or something?

File: Source/WebGPU/WebGPU/RenderPipeline.mm [hunk 4]
```diff
@@ -437,8 +437,16 @@ static void populateStencilOperation(MTLStencilDescriptor *mtlStencil, const WGP
     if (descriptor.vertex.bufferCount)
         mtlRenderPipelineDescriptor.vertexDescriptor = createVertexDescriptor(descriptor.vertex);
 
-    if (descriptor.primitive.nextInChain)
-        return RenderPipeline::createInvalid(*this);
+    MTLDepthClipMode mtlDepthClipMode = MTLDepthClipModeClip;
+    if (descriptor.primitive.nextInChain) {
+        if (descriptor.primitive.nextInChain->sType != WGPUSType_PrimitiveDepthClipControl)
```
Comment (litherum) [inline Source/WebGPU/WebGPU/RenderPipeline.mm]: `|| descriptor.primitive.nextInChain->nextInChain`

File: Source/WebGPU/WebGPU/RenderPipeline.mm [hunk 5]
```diff
@@ -450,19 +458,18 @@ static void populateStencilOperation(MTLStencilDescriptor *mtlStencil, const WGP
...
-
     MTLRenderPipelineReflection *reflection;
-    const auto& pipelineLayout = WebGPU::fromAPI(descriptor.layout);
-    bool hasBindGroups = pipelineLayout.numberOfBindGroupLayouts() > 0;
-    id<MTLRenderPipelineState> renderPipelineState = [m_device newRenderPipelineStateWithDescriptor:mtlRenderPipelineDescriptor options: hasBindGroups ? MTLPipelineOptionNone : MTLPipelineOptionArgumentInfo reflection:&reflection error:nil];
+    const PipelineLayout* pipelineLayout = nullptr;
+    if (descriptor.layout) {
+        if (auto& layout = WebGPU::fromAPI(descriptor.layout); layout.numberOfBindGroupLayouts())
+            pipelineLayout = &layout;
+    }
```
Comment (litherum) [inline Source/WebGPU/WebGPU/RenderPipeline.mm:466]: This seems unrelated

File: Websites/webkit.org/demos/webgpu/scripts/two-cubes.js [hunk 6]
```diff
@@ -182,7 +181,8 @@ async function helloCube() {
         fragment: fragmentStageDescriptor,
         primitive: {
             topology: "triangle-list",
-            cullMode: "back"
+            cullMode: "back",
+            unclippedDepth: true
```
Comment (litherum) [inline Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:187]: Don't we have to enable the feature at device creation time?

File: Source/WebGPU/WebGPU/RenderPipeline.mm [hunk 7]
```diff
@@ -437,8 +437,16 @@ static void populateStencilOperation(MTLStencilDescriptor *mtlStencil, const WGP
...
-        return RenderPipeline::createInvalid(*this);
+    MTLDepthClipMode mtlDepthClipMode = MTLDepthClipModeClip;
+    if (descriptor.primitive.nextInChain) {
+        if (descriptor.primitive.nextInChain->sType != WGPUSType_PrimitiveDepthClipControl)
+            return RenderPipeline::createInvalid(*this);
+
+        auto* depthClipControl = reinterpret_cast<const WGPUPrimitiveDepthClipControl*>(descriptor.primitive.nextInChain);
+
+        if (depthClipControl->unclippedDepth)
+            mtlDepthClipMode = MTLDepthClipModeClamp;
```
Comment (litherum) [inline Source/WebGPU/WebGPU/RenderPipeline.mm:448]: We don't have to do any workarounds or anything?

**Review by litherum — APPROVED** [review]

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp (hunk 2 above)
Comment (mwyrzykowski) [reply Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:484]: Good catch! I just discovered WebGPU requires opt-in for features

File: Source/WebGPU/WebGPU/RenderPipeline.mm (hunk 5 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/RenderPipeline.mm:466]: It is to avoid two `::create` functions with almost identical parameters

File: Source/WebGPU/WebGPU/RenderPipeline.h (hunk 3 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/RenderPipeline.h:46]: We could pass the `WGPURenderPipelineDescriptor` and let the constructor do the unpacking. What do you think of that idea?

File: Websites/webkit.org/demos/webgpu/scripts/two-cubes.js (hunk 6 above)
Comment (mwyrzykowski) [reply Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:187]: Yes that is correct. I will fix that

File: Source/WebGPU/WebGPU/RenderPipeline.mm (hunk 7 above)
Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/RenderPipeline.mm:448]: The extent of the workaround is the WGSL compiler should clamp fragDepth, when written, to the viewport z-bounds. I will file a bug for this

Comment (mwyrzykowski) [reply Source/WebGPU/WebGPU/RenderPipeline.mm:448]: https://bugs.webkit.org/show_bug.cgi?id=252729

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp (hunk 2 above)
Comment (mwyrzykowski) [reply Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:484]: I will add this validation to `Device::createRenderPipeline` so we return an invalid pipeline when the feature is not enabled and `unclippedDepth == true` which is what I believe the specification requires
